### PR TITLE
Rest API - User - Upload resume #464 & upload Avatar #459 & Project - Search By User #467

### PR DESCRIPTION
**Pull request by arpitsrm**
Rest API - User - Upload Avartar #459

### CODE REVIEW

**Comment by arpitsrm** [comment]
Not sure why 	Merge remote-tracking branch 'refs/remotes/Code4SocialGood/master' is committed? The actual changes for the tasks are in 
Rest API - User - Upload Avartar #459 
Upload resume #464  & upload Avartar #459 

Kindly have a look before merging

**Comment by Code4SocialGood** [comment]
```
 +	String AVARTAR_UPLOAD_DIRECTORY = "avartars";
 +	String RESUME_UPLOAD_DIRECTORY = "resumes";
 +    String AVARTAR_FORMAT = ".jpg";
 +    String RESUME_FORMAT = ".jpg";
```
1. Put the constants in a separate file.
2. Constants are defined as public static final.
3. Genealize the image file format as IMAGE_FORMAT, which convers avartar, project image, and organization log
4. Image accepts file formats beyong jpg, such as png and gif. Please study what image formats to accept
5. Resume format is not image, let's only allow pdf

```
 +import static org.c4sg.service.UserService.AVARTAR_UPLOAD_DIRECTORY;
 +import static org.c4sg.service.UserService.RESUME_UPLOAD_DIRECTORY;
```
1. Define as public contants, no need to import

**Comment by arpitsrm** [comment]
Ok.. Post Review changes are done.. It has changes for RestAPI - Organization uploadLogo : No File upload option #482 as well.
Please review and let me know if it fulfills the requirement.

File: C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.c4sg.constant;
+
+public class C4sgConstants {
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java]: Use a meaningful name to this constant

File: C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+	
+	public static final String AVARTAR_UPLOAD_DIRECTORY = "avartars";
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java]: change to AVATAR not avaRtar .

File: C4SGSvc/src/main/java/org/c4sg/controller/UserController.java [hunk 3]
```diff
@@ -81,4 +87,60 @@ public void deleteUser(@ApiParam(value = "User id to delete", required = true)
                                 @RequestParam(required = false) String lastName) {
         return userService.search(userName, firstName, lastName);
     }
+    
+    @RequestMapping(value = "/{userId}/uploadAvartar", method = RequestMethod.POST)
+    @ApiOperation(value = "Add new upload Avatar")
+    public String uploadAvartar(@ApiParam(value = "user Id", required = true)
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: change to AVATAR not avaRtar .

File: C4SGSvc/src/main/java/org/c4sg/controller/UserController.java [hunk 4]
```diff
@@ -81,4 +87,60 @@ public void deleteUser(@ApiParam(value = "User id to delete", required = true)
...
+                             @PathVariable Integer userId,
+                             @ApiParam(value = "Request Body", required = true)
+                             @RequestBody String avatarFileContent) {
+    	FileOutputStream fos=null;
+        try {
+            byte[] imageByte = Base64.decodeBase64(avatarFileContent);
+            File directory = new File(C4sgConstants.AVARTAR_UPLOAD_DIRECTORY);
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: change to AVATAR not avaRtar .

File: C4SGSvc/src/main/java/org/c4sg/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -90,4 +92,14 @@ public void deleteUser(Integer id) {
                 .map(p -> userMapper.getUserDtoFromEntity(p))
                 .collect(Collectors.toList());
     }
+
+	@Override
+	public String getAvatarUploadPath(Integer userId) {
+		return C4sgConstants.AVARTAR_UPLOAD_DIRECTORY + File.separator + userId + C4sgConstants.IMAGE_FORMAT;
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/service/impl/UserServiceImpl.java]: change to AVATAR not avaRtar .

File: C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java (hunk 1 above)
Comment (arpitsrm) [inline C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java]: How about ApplicationConstants?

Comment (Code4SocialGood) [inline C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java]: That's fine, one size fits all for now. We can split up to several constants when application grows.

Comment (Code4SocialGood) [inline C4SGSvc/src/main/java/org/c4sg/constant/C4sgConstants.java]: wait, see here, you need a more specific name for your constant class
image

**Comment by arpitsrm** [comment]
@felipealves-gnu : requested Changes have been addressed. Kindly review

File: C4SGSvc/src/main/java/org/c4sg/constant/ApplicationConstants.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package org.c4sg.constant;
+
+public class ApplicationConstants {
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/constant/ApplicationConstants.java]: Change for 2 **Enums** (follow the examples inside constants directory):
1) Called: Directory
    Variables: AVATAR_UPLOAD, RESUME_UPLOAD, etc
2) Called: Formats
    Variables: IMAGE, RESUME

File: C4SGSvc/src/main/java/org/c4sg/controller/UserController.java [hunk 7]
```diff
@@ -81,4 +87,60 @@ public void deleteUser(@ApiParam(value = "User id to delete", required = true)
...
+                             @RequestBody String avatarFileContent) {
+    	FileOutputStream fos=null;
+        try {
+            byte[] imageByte = Base64.decodeBase64(avatarFileContent);
+            File directory = new File(ApplicationConstants.AVATAR_UPLOAD_DIRECTORY);
+            if (!directory.exists()) {
+                directory.mkdir();
+            }
+            File f = new File(userService.getAvatarUploadPath(userId));
+            fos = new FileOutputStream(f);
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: Change to use Java 7 approach, follow this: https://www.mkyong.com/java/how-to-read-file-in-java-fileinputstream/

File: C4SGSvc/src/main/java/org/c4sg/service/impl/ProjectServiceImpl.java [hunk 8]
```diff
@@ -152,4 +153,32 @@ private void isUserAppliedPresent(Integer userId, Integer projectId) throws User
...
+	public List<ProjectDTO> findByUser(Integer userId) {
+		User user = userDAO.findById(userId);
+        requireNonNull(user, "Invalid User Id");
+		List<UserProject> userProjects = userProjectDAO.findByUserId(userId);
+		Comparator<UserProject> projectComp = new Comparator<UserProject>() {
+			@Override
+			public int compare(UserProject o1, UserProject o2) {
+				int result = 0;
+				result = o1.getStatus().compareTo(o2.getStatus());
+				//TODO Requirement skipped for now
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/service/impl/ProjectServiceImpl.java]: If it's not specified in the requirements delete those comments.

File: C4SGSvc/src/main/java/org/c4sg/util/FileUploadUtil.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+		 */
+		if (!(contentType.equals("image/pjpeg") || contentType.equals("image/jpeg") || contentType.equals("image/png")
+				|| contentType.equals("image/gif") || contentType.equals("image/bmp")
+				|| contentType.equals("image/x-png") || contentType.equals("type=image/x-icon"))) {
+			return false;
+		}
+		return true;
+	}
+
+	//TODO once logo upload functionality is decided, will work on resume upload
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/util/FileUploadUtil.java]: If it's not specified in the requirements delete those comments.

File: C4SGSvc/src/main/java/org/c4sg/controller/OrganizationController.java [hunk 10]
```diff
@@ -26,6 +27,41 @@
 
     @Autowired
     private OrganizationService organizationService;
+    
+    //Determining if this approach is better that using base64
+    @RequestMapping(value = "/{organizationId}/uploadLogoAsFile", method = RequestMethod.POST)
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/OrganizationController.java]: Change from organizationId to just id

File: C4SGSvc/src/main/java/org/c4sg/controller/ProjectController.java [hunk 11]
```diff
@@ -92,6 +92,28 @@ public Project getProject(@ApiParam(value = "Name of project to return", require
     }
     
     @CrossOrigin
+    @RequestMapping(value = "/search/byUser/{userId}", method = RequestMethod.GET)
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/ProjectController.java]: Change from userId to just id

File: C4SGSvc/src/main/java/org/c4sg/controller/UserController.java [hunk 12]
```diff
@@ -81,4 +87,60 @@ public void deleteUser(@ApiParam(value = "User id to delete", required = true)
                                 @RequestParam(required = false) String lastName) {
         return userService.search(userName, firstName, lastName);
     }
+    
+    @RequestMapping(value = "/{userId}/uploadAvatar", method = RequestMethod.POST)
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: Change from userId to just id

File: C4SGSvc/src/main/java/org/c4sg/controller/UserController.java [hunk 13]
```diff
@@ -81,4 +87,60 @@ public void deleteUser(@ApiParam(value = "User id to delete", required = true)
...
+    @ApiOperation(value = "Add new upload Avatar")
+    public String uploadAvatar(@ApiParam(value = "user Id", required = true)
+                             @PathVariable Integer userId,
+                             @ApiParam(value = "Request Body", required = true)
+                             @RequestBody String avatarFileContent) {
```
Comment (felipealves-gnu) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: Don't we need a  `@RequestPart("file") MultipartFile avatarFileContent` , here ?

Comment (arpitsrm) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: Yes.. we do need that here.. But I had a discussion regarding it with Jenny and she wanted to get it tested by frontEnd developers for uploadLogo
I created the task for the same "RestAPI - Organization uploadLogo : No File upload option #482"

Comment (Code4SocialGood) [inline C4SGSvc/src/main/java/org/c4sg/controller/UserController.java]: I haven't got a frontend developer signed up on it yet.

**Comment by arpitsrm** [comment]
@Code4SocialGood , @felipealves-gnu 
Review changes done

**Pull request closed without merging**